src/pybind/mgr/rgw/module.py

```diff
@@ -99,25 +103,78 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
                                  zone_name: Optional[str] = None,
                                  endpoints: Optional[str] = None,
                                  sys_uid: Optional[str] = None,
                                  uid: Optional[str] = None,
-                                 start_radosgw: Optional[bool] = True):
+                                 start_radosgw: Optional[bool] = True,
+                                 zonegroup_is_master: Optional[bool] = True,
+                                 zone_is_master: Optional[bool] = True):
         """Bootstrap new rgw realm, zonegroup, and zone"""
-
-
         try:
-            retval, out, err = RGWAM(self.env).realm_bootstrap(realm_name, zonegroup_name,
-                    zone_name, endpoints, sys_uid, uid, start_radosgw)
+            retval, out, err = RGWAM(self.env).realm_bootstrap(realm_name,
+                                                               zonegroup_name,
+                                                               zone_name,
+                                                               endpoints,
+                                                               sys_uid,
+                                                               uid,
+                                                               start_radosgw,
+                                                               zonegroup_is_master,
+                                                               zone_is_master)
         except RGWAMException as e:
             self.log.error('cmd run exception: (%d) %s' % (e.retcode, e.message))
             return (e.retcode, e.message, e.stderr)
 
         return HandleCommandResult(retval=retval, stdout=out, stderr=err)
 
+    @CLICommand('rgw apply', perm='rw')
+    def _cmd_rgw_apply(self, inbuf: Optional[str] = None):
+        """apply rgw spec"""
+
+        usage = """Usage: ceph rgw apply -i <yaml spec>"""
+        if not inbuf:
+            return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr=usage)
+
+        # YAML '---' document separator with no content generates
+        # None entries in the output. Let's skip them silently.
+        yaml_objs: Iterator = yaml.safe_load_all(inbuf)
+        specs = [o for o in yaml_objs if o is not None]
+        outs = []
+        err = {}
+        retval = 0
+        for spec in specs:
+            try:
+                realm_name = spec.get('realm_name')
+                zonegroup_name = spec.get('zonegroup_name')
+                zone_name = spec.get('zone_name')
+                if not all([realm_name, zonegroup_name, zone_name]):
+                    err = {'error': 'spec fields: realm_name, zonegroup_name and zone_name are mandatory',
+                           'retcode': errno.EINVAL}
+                    return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr=json.dumps(err))
+                retval, out, err_msg = RGWAM(self.env).realm_bootstrap(realm_name,
```

Comment:
I feel like it would be worth having the option to specify a placement in the spec here if you want the orchestrator to deploy the rgw daemons. Otherwise, we're just going to place them on random nodes and users will have to go manually edit the spec the orchestrator has to say where they actually want them.

Reply:
New code uses cephadm RGWSpec directly this way we can use the same params already supported by cephadm when specifying RGW params (including placement).